### PR TITLE
Bugfix/deep environment variables

### PR DESCRIPTION
**Pull request by emilesalem**
the fix that was added for issue #3944 is not evaluating environment variables beyond the first level of hash  values, as described [here](https://discuss.elastic.co/t/bug-in-http-poller-environment-variables-dont-always-evaluate-with-http-poller/53070)

This pull request aims to correct that. 

I could not add a test as I dont know anything about Ruby development and failed to run the test suites. I tested this fix running logstash from source inside a docker image based on jruby:9.1-jdk and it works with a few sample configurations. I didnt have the same success trying to run test suites. 

Thank you for considering this fix.

### CODE REVIEW

**Comment by emilesalem** [comment]
I have now signed the CLA.

**Comment by z-vr** [comment]
thanks mate, that's exactly the problem I've been having, that solved it 👍

File: logstash-core/lib/logstash/config/mixin.rb [hunk 1]
```diff
@@ -47,15 +47,32 @@ def self.included(base)
...
   end
 
+  # Recursive method to replace environment variable references in parameters
+  def deep_replace(value)
+    if (value.is_a?(Hash))
+      value.each do |valueHashKey, valueHashValue|
+          value[valueHashKey.to_s] = deep_replace(valueHashValue)
+      end
+    else
+        if (value.is_a?(Array))
```
Comment (andrewvc) [inline logstash-core/lib/logstash/config/mixin.rb]: Please use 2-space indenting as per the rest of this project.

File: logstash-core/lib/logstash/config/mixin.rb [hunk 2]
```diff
@@ -47,15 +47,32 @@ def self.included(base)
...
+           value.each_index do | valueArrayIndex|
+                value[valueArrayIndex] = deep_replace(value[valueArrayIndex])
+                end
```
Comment (andrewvc) [inline logstash-core/lib/logstash/config/mixin.rb]: Indentation is off.

File: logstash-core/lib/logstash/config/mixin.rb [hunk 3]
```diff
@@ -47,15 +47,32 @@ def self.included(base)
...
+        else
+            return replace_env_placeholders(value)
+        end
+    end
+  end
+
   def config_init(params)
     # Validation will modify the values inside params if necessary.
     # For example: converting a string to a number, etc.
-
```
Comment (andrewvc) [inline logstash-core/lib/logstash/config/mixin.rb]: Please remove these unnecessary whitespace changes added by your editor.

File: logstash-core/lib/logstash/config/mixin.rb [hunk 4]
```diff
@@ -88,7 +105,7 @@ def config_init(params)
       next if params.include?(name.to_s)
       if opts.include?(:default) and (name.is_a?(Symbol) or name.is_a?(String))
         # default values should be cloned if possible
-        # cloning prevents
```
Comment (andrewvc) [inline logstash-core/lib/logstash/config/mixin.rb]: Please undo this whitespace change.

File: logstash-core/lib/logstash/config/mixin.rb [hunk 5]
```diff
@@ -105,22 +122,9 @@ def config_init(params)
...
-      else
-        if (value.is_a?(Array))
-          value.each_index do |valueArrayIndex|
-            value[valueArrayIndex] = replace_env_placeholders(value[valueArrayIndex])
-          end
-        else
-          params[name.to_s] = replace_env_placeholders(value)
-        end
-      end
+      params[name.to_s] = deep_replace(value)
```
Comment (andrewvc) [inline logstash-core/lib/logstash/config/mixin.rb]: Glad to see this pulled out into its own method :)

**Comment by suyograo** [comment]
@emilesalem can you update this PR with the requested changes?

**Comment by emilesalem** [comment]
I will try to have those changes done by the end of the week.

**Comment by emilesalem** [comment]
I have updated the pull request to correct the indentation errors. Unfortunately, as I explained in my first comment, I know basically nothing of Ruby development and do not have the time to learn it properly, so writing tests is beyond my abilities right now. Someone else will have to write the specs, I'm sorry.

**Comment by suyograo** [comment]
@emilesalem I'll add tests to it. Can you please open the PR against master instead of 5.1? Typically all patches must be submitted against the master branch and we then backport to release branches such as 5.1, 5.2 etc.

**Comment by karmi** [comment]
Hi @emilesalem, we have found your signature in our records, but it seems like you have signed with a different e-mail than the one used in yout Git [commit](https://github.com/elastic/logstash/pull/6392.patch). Can you please add both of these e-mails into your Github profile (they can be hidden), so we can match your e-mails to your Github profile?

**Comment by karmi** [comment]
Hi @emilesalem, we have found your signature in our records, but it seems like you have signed with a different e-mail than the one used in yout Git [commit](https://github.com/elastic/logstash/pull/6392.patch). Can you please add both of these e-mails into your Github profile (they can be hidden), so we can match your e-mails to your Github profile?

**Comment by z-vr** [comment]
has this not been released?

**Comment by emilesalem** [comment]
this was moved to https://github.com/elastic/logstash/pull/6694

merged in master 3 weeks ago. 

No idea when they'll release.
